Apply working draft content to status codes (#762)
zalando/restful-api-guidelines: chapters/http-status-codes-and-errors.adoc

[[status-code-207]]
==== 207 Multi-Status {rfc-status-207} {use} {document} {POST} ({DELETE})
[.indent]
The response body contains status information for multiple different parts of a
batch/bulk request (see <<152>> for details). Normally used for {POST}, in some
cases also for {DELETE}.

[[redirection-codes]]
=== Redirection codes

See <<250>>.

[.do-not-use]
[[status-code-301]]
==== 301 Moved Permanently {rfc-status-301} {do-not-use} {ALL}
[.indent]
This and all future requests should be directed to the
given URI.

[.do-not-use]
[[status-code-302]]
==== 302 Found {rfc-status-302} {do-not-use} {ALL}
[.indent]
This is a temporary redirect where some clients {MAY} change the request method
from {POST} to {GET}. Mainly used for dismissing and redirecting form
submissions in browsers. 

[.do-not-use]
[[status-code-303]]
==== 303 See Other {rfc-status-303} {do-not-use} {POST} {PUT} {PATCH} {DELETE}
[.indent]
The response to the request can be found under another URI using a
{GET} method. A disambiguated version of {302} for the case where the client
{MUST} change the method to {GET}.

[[status-code-304]]
==== 304 Not Modified {rfc-status-304} {document} {GET} {HEAD}
[.indent]

> Indicates that a conditional {GET} or {HEAD} request would have resulted in
> {200} response if it were not for the fact that the condition evaluated to false
> i.e. resource has not been modified since the date or version passed via request
> headers {If-Modified-Since} or {If-None-Match}.

Reviewer: Should we link here (304) also to 412 (the corresponding answer for conditional POST/PUT/DELETE requests)?

```suggestion
Indicates that a conditional {GET} or {HEAD} request would have resulted in
{200} response if it were not for the fact that the condition evaluated to false
i.e. resource has not been modified since the date or version passed via request
headers {If-Modified-Since} or {If-None-Match}.

Note that for conditional requests which _do_ change something (POST/PUT/PATCH/DELETE), you would use {412} if the condition failed.
```

Author: Good idea! Added a commit with additions to 304 and 412 with a format somewhere in between the suggestions :)

[[status-code-400]]
==== 400 Bad Request {rfc-status-400} {use} {document} {ALL}
[.indent]
Unspecific client error indicating that the server cannot process the request
due to something that is perceived to be a client error (e.g. malformed request
syntax, invalid request). Should also be delivered in case of input payload
fails business logic / semantic validation (instead of using {422}).

[[status-code-401]]
==== 401 Unauthorized {rfc-status-401} {use} {do-not-document} {ALL}
[.indent]
Actually *Unauthenticated*. The credentials are missing or not valid for the
target resource. For an API, this usually means that the provided token or
cookie is not valid. As this can happen for almost every endpoint, APIs should
normally not document this.

[[status-code-403]]
==== 403 Forbidden {rfc-status-403} {do-not-document} {ALL}
[.indent]
The user is not authorized to use this resource. For an API, this can mean that
the request's token was valid, but was missing a scope for this endpoint. Or
that some object-specific authorization failed. We recommend only documenting
the second case.

[[status-code-404]]
==== 404 Not found {rfc-status-404} {do-not-document} {ALL}
[.indent]

> The target resource was not found. This will be returned by most paths on most

Reviewer: ```suggestion
The target resource was not found. This will be returned by most (not documented) paths on most
```

APIs, and for endpoints with parameters when those parameters don't map to an
existing entity. For a {PUT} endpoint which only supports updating existing
resources, this might be returned if the resource does not exist. Apart from
these special cases, this does not need to be documented.

[[status-code-405]]
==== 405 Method Not Allowed {rfc-status-405} {document} {ALL}
[.indent]
The request method is not supported for this resource. In theory, this can be
returned for all resources for all the methods except the ones documented. Using
this response code for an existing endpoint (usually with path parameter) only
makes sense if it depends on some internal resource state whether a specific
method is allowed, e.g. an order can only be canceled via {DELETE} until the
shipment leaves the warehouse. *Do not use it unless you have such a special use
case, but then make sure to document it, making it clear why a resource might
not support a method.*

[[status-code-406]]
==== 406 Not Acceptable {rfc-status-406} {do-not-document} {ALL}
[.indent]

> Resource only supports generating content that is not listed in the {Accept}
> header sent in the request.

Reviewer: ```suggestion
Resource only supports generating content with content-types that are not listed
in the {Accept} header sent in the request.
```

return {200} or {204} and not {409}, even if the resource exists already. If any
`If-*` headers cause a conflict, you should use {412} and not {409}. Only
applicable to methods which change something.

[[status-code-410]]
==== 410 Gone {rfc-status-410} {do-not-document} {ALL}
[.indent]
The resource does not exist any longer (but did exist in the past), and will
most likely not exist in the future. This can be used e.g. when accessing a
resource that has intentionally been deleted. This normally does not need to be
documented, unless there is a specific need to distinguish this case from the
normal {404}.

[[status-code-411]]
==== 411 Length Required {rfc-status-411} {document} {POST} {PUT} {PATCH}
[.indent]
The server requires a {Content-Length} header for this request. This is normally
only relevant for large media uploads. The corresponding header parameter should
be marked as required. If used, it {MUST} to be documented (and explained). Only
applicable for methods with a request body.

[[status-code-412]]
==== 412 Precondition Failed {rfc-status-412} {do-not-document} {PUT} {PATCH} {DELETE}
[.indent]

> Returned for conditional requests, e.g. {If-Match} if the condition failed. Used
> for optimistic locking. Normally only applicable to methods that change
> something.

Reviewer: ```suggestion
Returned for conditional requests, e.g. {If-Match} if the condition failed. Used
for optimistic locking. Normally only applicable to methods that change
something. For HEAD/GET, use {304} instead.
```

```suggestion
Returned for conditional requests, e.g. {If-Match} if the condition failed. Used
for optimistic locking. Normally only applicable to methods that change
something. See also {304}.
```

Author: See above: https://github.com/zalando/restful-api-guidelines/pull/762#discussion_r1187168491

[[status-code-415]]
==== 415 Unsupported Media Type {rfc-status-415} {do-not-document} {POST} {PUT} {PATCH}
[.indent]

> The client did not provide a supported request body. Only applicable to methods
> with a request body.

Reviewer: ```suggestion
The client did not provide a supported content-type for the request body.
Only applicable to methods with a request body.
```

[.do-not-use]
[[status-code-417]]

> ==== 417 Locked {rfc-status-417} {do-not-use} {ALL}

Reviewer: ```suggestion
==== 417 Expectation Failed {rfc-status-417} {do-not-use} {ALL}
```

Looks like a copy+paste mistake.

[[status-code-428]]
==== 428 Precondition Required {rfc-status-428} {do-not-document} {ALL}
[.indent]

> Server requires the request to be conditional, e.g. to make sure that the "lost
> update problem" is avoided (see <<181>>).

Reviewer: This seems to link to the wrong rule.
```suggestion
Server requires the request to be conditional, e.g. to make sure that the "lost
update problem" is avoided (see <<182>>). Instead of documenting this
response status, the required headers should be documented (and marked
as required).
```
